src/components/inline_edit/inline_edit_form.tsx

```diff
@@ -0,0 +1,330 @@
+  const setReadModeRefs = useCombinedRefs([
+    readModeFocusRef,
+    readModeProps?.buttonRef,
+  ]);
+  const setEditModeRefs = useCombinedRefs([
+    editModeFocusRef,
+    editModeProps?.inputProps?.inputRef,
+  ]);
+
+  const [isEditing, setIsEditing] = useState(false || startWithEditOpen);
+  const [editModeValue, setEditModeValue] = useState(defaultValue);
+  const [readModeValue, setReadModeValue] = useState(defaultValue);
+
+  const activateEditMode = () => {
+    setIsEditing(true);
+    // Waits a tick for state to settle and the focus target to render
+    requestAnimationFrame(() => editModeFocusRef.current?.focus());
+  };
+
+  const cancelInlineEdit = () => {
+    setEditModeValue(readModeValue);
+    setIsEditing(false);
+    requestAnimationFrame(() => readModeFocusRef.current?.focus());
+  };
+
+  const saveInlineEditValue = async () => {
+    // If an onSave callback is present, and returns false, stay in edit mode
+    if (onSave) {
+      const onSaveReturn = onSave(editModeValue);
+      const awaitedReturn =
+        onSaveReturn instanceof Promise ? await onSaveReturn : onSaveReturn;
+      if (awaitedReturn === false) return;
+    }
+
+    setReadModeValue(editModeValue);
+    setIsEditing(false);
+    requestAnimationFrame(() => readModeFocusRef.current?.focus());
+  };
+
+  const editModeInputOnKeyDown = (event: KeyboardEvent<HTMLElement>) => {
+    switch (event.key) {
+      case keys.ENTER:
+        saveInlineEditValue();
+        break;
+      case keys.ESCAPE:
+        cancelInlineEdit();
+        break;
+    }
+  };
+
+  const editModeForm = (
+    <EuiFlexGroup gutterSize="s">
+      <EuiFlexItem>
+        <EuiFormRow
+          fullWidth
+          isInvalid={isInvalid}
+          error={isInvalid && editModeProps?.formRowProps?.error}
+          {...editModeProps?.formRowProps}
+        >
+          <EuiFieldText
+            fullWidth
+            value={editModeValue}
+            aria-label={inputAriaLabel}
+            compressed={sizes.compressed}
+            isInvalid={isInvalid}
+            isLoading={isLoading}
+            data-test-subj="euiInlineEditModeInput"
+            {...editModeProps?.inputProps}
+            inputRef={setEditModeRefs}
+            onChange={(e) => {
+              setEditModeValue(e.target.value);
+              editModeProps?.inputProps?.onChange?.(e);
+            }}
+            onKeyDown={(e) => {
+              editModeInputOnKeyDown(e);
+              editModeProps?.inputProps?.onKeyDown?.(e);
+            }}
+            aria-describedby={classNames(
+              editModeDescribedById,
+              editModeProps?.inputProps?.['aria-describedby']
+            )}
+          />
+        </EuiFormRow>
+        <span id={editModeDescribedById} hidden>
+          <EuiI18n
+            token="euiInlineEditForm.inputKeyboardInstructions"
+            default="Press Enter to save your edited text. Press Escape to cancel your edit."
```

Comment:
I'm really on the fence on this one because of how different screen readers (and even the same screen reader with different browsers behave:

* **Safari + VO:** `Enter` does not do anything. `Esc` cancels and switches to read mode.
* **NVDA + Chrome:** `Enter` switches into our out of Focus mode. `Esc` cancels and switches to read mode.
* **NVDA + Firefox:** `Enter` saves and switches to read mode. `Esc` cancels and switches to read mode.

We should definitely keep the `Enter` and `Esc` shortcuts, but I wonder if the directions could say something like

```suggestion
            default="Press Tab when you are done editing to save or cancel your edit."
```

I'm not sold on this language because it obfuscates the behaviors and asks more of users in discovery. I think I'd rather have users need to figure out a slightly longer (albeit consistent) path, than one that has such different behaviors.

What do you think?

Reply:
My mediumly-strong opinion is that rather than tell SR users to tab to save, I would rather we just fix the `Enter` behavior in Safari

Reply:
I can roll with Cee's opinion to look for a fix to the Safari `Enter` keypress behavior, rather than change the language. We don't have real-world feedback about the UX of different language / UI behavior combinations (yet). I'll update my review.